controller/eni_ip_controller.go

```diff
@@ -0,0 +1,267 @@
+		// Every object will trigger the `Updatefunc` even if there have been no actual updates triggered.
+		// In some cases you can set this to a very high interval - as you can assume you will see periodic updates in normal operation.
+		// The interval is set low here for demo purposes.
+		10*time.Second,
+		// Callback Functions to trigger on add/update/delete
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if key, err := cache.MetaNamespaceKeyFunc(obj); err == nil {
+					log.Infof("Add key %s", key)
+					eniIPController.queue.Add(key)
+				}
+			},
+			/*
+				// TODO feedback from node about their eni requirement
+					UpdateFunc: func(old, new interface{}) {
+						if key, err := cache.MetaNamespaceKeyFunc(new); err == nil {
+							log.Infof("Update key %s", key)
+							//eniIPController.queue.Add(key)
+						}
+					},
+			*/
+			DeleteFunc: func(obj interface{}) {
+				if key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj); err == nil {
+					log.Infof("Delete key %s", key)
+					eniIPController.queue.Add(key)
+				}
+			},
+		},
+		cache.Indexers{},
+	)
+
+	eniIPController.informer = informer
+	eniIPController.indexer = indexer
+	eniIPController.nodeLister = lister_v1.NewNodeLister(indexer)
+	eniIPController.nodeStoreSynced = informer.HasSynced
+
+	return eniIPController
+}
+
+func (c *ENIIPController) processNextItem() bool {
+	// Wait until there is a new item in the working queue
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// Tell the queue that we are done with processing this key. This unblocks the key for other workers
+	// This allows safe parallel processing because two pods with the same key are never processed in
+	// parallel.
+	defer c.queue.Done(key)
+
+	// Invoke the method containing the business logic
+	err := c.nodeENIIPController(key.(string))
+	// Handle the error if something went wrong during the execution of the business logic
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *ENIIPController) nodeENIIPController(key string) error {
+	log.Infof("Processing: %s", key)
+	obj, exists, err := c.indexer.GetByKey(key)
+	if err != nil {
+		log.Errorf("Fetching object with key %s from store failed with %v", key, err)
+		return err
+	}
+	if !exists {
+		log.Infof("Node %s does not exist anymore", key)
+		return nil
+	}
+
+	node, ok := obj.(*v1.Node)
+	if !ok {
+		log.Errorf("Object with key %s [%v] is not a node!", key, obj)
+		//TODO err
+		return nil
+	}
+
+	instanceType, ok := node.Labels[kubeletapis.LabelInstanceType]
+	if !ok {
+		log.Infof("Node %s has no labels %s", key, instanceTypeLabel)
+		// TODO query EC2 about instance type
+		return nil
+	}
+
+	ipLimit, err := vpcipresource.GetVPICIPLimit(instanceType)
+
+	if err != nil {
+		log.Errorf("Failed to get vpc ip limit %v", err)
+		//TODO handle err
+		return nil
+	}
+
+	if strings.Compare(key, c.nodeName) == 0 {
```

Comment:
This seems wrong.  If it is a running tally of IPs, then we should be doing it for all pods that are already running on the node.  But its called IP limit, so that means to me it should be a limit, and should have nothing to do with any pods currently running on the instance.

Reply:
This assumes eni-controller pod is deployed before any other workload Pods and all other Pods have specified using resource of ***vpc.amazonaws.com/ipv4*** resources.